time_utils/automation/lib/python/community/time_utils.py

```diff
@@ -108,45 +115,99 @@ def to_datetime(when, log=logging.getLogger("{}.time_utils".format(LOG_PREFIX)))
+            dt_python = to_python_datetime(dt_java)
+
         elif isinstance(when, int):
+            log.debug('when is int')
             dt = DateTime().now().plusMillis(when)
-        elif isinstance(when, (scope.DateTimeType)):
+            dt_java = ZonedDateTime.now().plus(when, ChronoUnit.MILLIS)
+            dt_python = to_python_datetime(dt_java)
+
+        elif isinstance(when, DateTimeType):
+            log.debug('when is DateTimeType')
             dt = DateTime(str(when))
-        elif isinstance(when, (scope.DecimalType, scope.PercentType,
-                            scope.QuantityType)) :
+            dt_java = to_java_zoneddatetime(dt)
+            dt_python = to_python_datetime(dt_java)
+
+        elif isinstance(when, (DecimalType, PercentType, QuantityType)):
+            log.debug('when is decimal, percent or quantity type')
             dt = DateTime().now().plusMillis(when.intValue())
+            dt_python = datetime.now() + timedelta(milliseconds = when.intValue())
+            dt_java = to_java_zoneddatetime(dt_python)
+
         elif isinstance(when, datetime):
-            dt = to_joda_datetime(when)
+            log.debug('when is datetime')
+            dt_python = when
+            dt_java = to_java_zoneddatetime(when)
+            #get system time zone to avoid convertion errors
+            dt = to_joda_datetime(when.replace(tzinfo=None)) 
+
+        elif isinstance(when, time):
+            log.debug('when is python time object')
+            dt_java = ZonedDateTime.now().withHour(when.hour).withMinute(when.minute).withSecond(when.second)
+            dt_python = to_python_datetime(dt_java)
+            dt = to_joda_datetime(dt_java)
+
+        elif isinstance(when, (str, unicode)):
+            if is_iso8601(when):
+                log.info('when is iso8601: '+str(when))
+                dt = DateTime(when)
+                dt_python = datetime.strptime(str(when), "%Y-%m-%dT%H:%M:%S.%fffff%Z")
+                #get system time zone to avoid convertion errors
+                dt_java = to_java_zoneddatetime(dt_python.replace(tzinfo=None))
+            else:
+                log.debug('when is duration')
+                log.debug(str(when))
+                dt = parse_duration_to_datetime(when, log)
+                log.debug('dt is ' + str(dt))
+                dt_python = datetime.now() + parse_duration(when, log)
+                log.debug('dt python is ' + str(dt))
+                #get system time zone to avoid convertion errors
+                dt_java = to_java_zoneddatetime(dt_python.replace(tzinfo=None))
         else:
-            log.warn("When is an unknown type {}".format(type(when)))
+            log.warn('When is an unknown type {}'.format(type(when)))
     except:
-        import traceback
-        log.error("Exception: {}".format(traceback.format_exc()))
-    finally:
+        log.error('Exception: {}'.format(traceback.format_exc()))
+
+    if output == 'python':
```

Comment:
I think there is a better way to structure this rule overall to be more efficient. Rather than creating the joda, zoned, and python time for each and every type and then throwing two of those away at the end, just calculate one of them, probably ZonedDateTime. Then at the end, only create the Joda or Python DateTime if it's called for. That should reduce the lines of code and avoid throw/away work.